api/pkg/s3/object-utils.go

```diff
@@ -0,0 +1,100 @@
+/*
+ * Minio Cloud Storage, (C) 2015, 2016 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package s3
+
+import (
+	"regexp"
+	"strconv"
+	"strings"
+	"unicode/utf8"
+	"fmt"
+)
+
+// validBucket regexp.
+var validBucket = regexp.MustCompile(`^[a-z0-9][a-z0-9\.\-]{1,61}[a-z0-9]$`)
+
+// IsValidBucketName verifies a bucket name in accordance with Amazon's
+// requirements. It must be 3-63 characters long, can contain dashes
+// and periods, but must begin and end with a lowercase letter or a number.
+// See: http://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html
+func isValidBucketName(bucketName string) bool {
```

Comment:
why do we put Amazon's validity logic in our code...is there no API we can call from the Amazon SDK to validate the bucket name ?
My concern is how will we keep this in sync when Amazon changes their bucket name validation rules ?

Reply:
@rhsakarpos the function isValidBucketName here is migrated from yig. and because api layer provided by s3 in multi-cloud must be compatible with s3 protocol, and this function is implemented according to s3 protocol, so, this function works fine here.

Reply:
sorry, i do not agree with this. imho, we should NOT have any cloud vendor specific code in OSDS

Reply:
Hi @rhsakarpos , imho, this check is definitely OSDS check, but as AWS S3 is de facto standard, and we want to be compatible with it, so we did check like this.

Reply:
hi @sfzeng imho, we do NOT need any check (with code) in OSDS. We simply call the cloud provider API...S3 or Azure or whatever cloud backend.

Reply:
hi @rhsakarpos , this bucket is not the one created in AWS or Azure, or any other cloud providers. It's OSDS bucket. Say, we have an AWS bucket aws_bucket, and register it as a backend aws_backend in OSDS, then we can create bucket bucket1 in OSDS. what we check is bucket1, not aws_bucket.